src/aspire/source/micrograph.py

```diff
@@ -1,75 +1,389 @@
+            Image(micrograph).save(mrc_file_path, overwrite=overwrite)
+
+            output_mrc_filenames.append(mrc_file_path)
+
+        return output_mrc_filenames
+
+    def asnumpy(self):
+        """
+        Return micrograph data as dense Numpy array.
+
+        :return: Numpy array.
+        """
+        return self.images[:]._data
+
+    def show(self, *args, **kwargs):
+        """
+        Helper function to display micrograph. See Image.show().
+        """
+        Image(self.asnumpy()).show(*args, **kwargs)
+
+    @property
+    def images(self):
+        """
+        Returns micrographs.
+
+        :return: An `ImageAccessor` for the noisy micrographs.
+        """
+        return self._images_accessor
+
+    @abstractmethod
+    def _images(self, indices):
+        """
+        Accesses and returns micrographs.
+
+        :param indices: A 1-D Numpy array of integer indices.
+        :return: An `ArrayMicrographSource` object representing the micrographs for `indices`.
+        """
+
+
+class ArrayMicrographSource(MicrographSource):
+    def __init__(self, micrographs, dtype=None):
+        """
+        Instantiate a `MicrographSource` with `micrographs`.
+
+        :param micrographs: Numpy array (count, L, L)
+            where  `L` represents the size in pixels.
+        :param dtype: Data type of returned `MicrographSource`.
+            Default of `None` will infer dtype from the array.
+            Explicitly setting dtype will attempt a cast.
+            Currently only `float32` and `float64` are supported.
+            Note, due to limitations of common MRC implementations,
+            saving is limited to single precision.
+        """
+
+        # Check micrographs is an array
+        if not isinstance(micrographs, np.ndarray):
+            raise NotImplementedError(
+                f"{self.__class__.__name__} not implemented for {type(micrographs)}."
+            )
+
+        # Ensure dimensions
+        if micrographs.ndim == 2:
+            micrographs = micrographs[None, :, :]
+
+        if micrographs.ndim != 3 or (micrographs.shape[-2] != micrographs.shape[-1]):
+            raise NotImplementedError(
+                f"Incompatible `micrographs` shape {micrographs.shape}, expects (count, L, L)"
+            )
+
+        super().__init__(
+            micrograph_count=micrographs.shape[0],
+            micrograph_size=micrographs.shape[-1],
+            dtype=dtype or micrographs.dtype,
+        )
+
+        # We're already backed by an array, access it directly.
+        self._data = micrographs.astype(self.dtype, copy=False)
+
+    def _images(self, indices):
+        """
+        Accesses and returns micrographs from a Numpy array.
+
+        :param indices: A 1-D Numpy array of integer indices.
+        :return: An array backed `MicrographSource` object representing the micrographs for `indices`.
+        """
+        return ArrayMicrographSource(self._data[indices])
```

Comment:
I'm a bit confused here. Doesn't this mean that calling `mg_src.images[k]` will return another `MicrographSource` (containing a single micrograph)? Don't we want it to return an `Image`?

Reply:
Could go either way.  Here both `ArrayMicrographSource` and `DiskMicrographSource` returned `ArrayMicrographSource`.  For some reason it made more sense to me for a MicrographSource to return a Micrograph at the time.

 I just changed towards `Image` and updated the docstring. Other than `type` of the object it probably doesn't matter to the end user. If a users wants to dress it up with metadata they can wrap it back into a `MicrographSource` on their own.